### PR TITLE
Fix STJ source gen support for System.BinaryData.

### LINKED ISSUE #81702: BinaryData + JSON source gen produces compiler warnings
EDIT by @eiriktsarpalis to reflect more recent error modes

In https://github.com/dotnet/runtime/pull/73448 we've added internal converter for BinaryData but that's not working with JSON source-gen.

## Repro

```csharp
using System.Text.Json.Serialization;

[JsonSerializable(typeof(BinaryData))]
partial class MyContext : JsonSerializerContext { }
```

## Expected behavior:
Compiles and works when you serialize/deserialize

## Actual

```
warning SYSLIB1041: The 'JsonConverterAttribute' type 'System.BinaryDataConverter' specified on member 'System.BinaryData' is not a converter type or does not contain an accessible parameterless constructor.
```

## API Proposal

We should make the constructor of the already included custom converter public:

```diff
namespace System;

[JsonConverter(typeof(BinaryDataConverter))]
public partial class BinaryData { }

-internal sealed class BinaryDataConverter : JsonConverter<BinaryData> { }
+namespace System.Text.Json.Serialization;
+
+public sealed class BinaryDataConverter : JsonConverter<BinaryData> { }
```

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix https://github.com/dotnet/runtime/issues/81702. Also fixes an issue I discovered where BOM trimming was not being done in the `ToObjectFromJson` overload accepting `JsonTypeInfo<T>`.

### CODE REVIEW

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

Nice catch on the BOM issue.

File: src/libraries/System.Memory.Data/src/System/BinaryData.cs [hunk 1]
```diff
@@ -310,7 +302,18 @@ public override unsafe string ToString()
...
-            => JsonSerializer.Deserialize<T>(_bytes.Span, jsonTypeInfo);
+            => JsonSerializer.Deserialize(GetBytesWithTrimmedBom(), jsonTypeInfo);
+
+        private ReadOnlySpan<byte> GetBytesWithTrimmedBom()
+        {
+            ReadOnlySpan<byte> span = _bytes.Span;
+
+            // Check for the UTF-8 byte order mark (BOM) EF BB BF
+            if (span.Length > 2 && span[0] == 0xEF && span[1] == 0xBB && span[2] == 0xBF)
+                span = span.Slice(3);
```
Comment (Sergio0694) [inline src/libraries/System.Memory.Data/src/System/BinaryData.cs:313]: Might want to use a pattern if you're feeling particularly fancy? 😄
```suggestion
            if (span is [0xEF, 0xBB, 0xBF, ..ReadOnlySpan<byte> tail])
                span = tail;
```

Comment (eiriktsarpalis) [reply src/libraries/System.Memory.Data/src/System/BinaryData.cs:313]: The library also targets netstandard and we can't use polysharp unfortunately :-(

File: src/libraries/System.Memory.Data/src/CompatibilitySuppressions.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Memory.Data/src/CompatibilitySuppressions.xml:1]: @ViktorHofer just to confirm is this the appropriate location for package apicompat suppresions? It was just placed there by the generator.

Comment (ViktorHofer) [reply src/libraries/System.Memory.Data/src/CompatibilitySuppressions.xml:1]: Yes, that's the correct location. This change was technically non breaking but APICompat filed it as the current attributes rule checks for equality. It doesn't know that replacing the Type in the JsonConverterAttribute shouldn't be breaking.

**Pull request merged**